capi-configuration/README.md

```diff
@@ -0,0 +1,333 @@
+      infrastructureRef:
+        name: capa-mp-0
+```
+
+This minimal custom resource can also be represented as a template which will remain valid throughout the lifecycle of `v1alpha3` in this case:
+```yaml
+apiVersion: cluster.x-k8s.io/v1alpha3
+kind: MachinePool
+metadata:
+  name: {{clusterID}}-mp-0
+spec:
+  clusterName: {{clusterID}}
+  template:
+    spec:
+      bootstrap:
+        configRef:
+          name: {{clusterID}}-mp-0
+      infrastructureRef:
+        name: {{clusterID}}-mp-0
+```
+This template is going to be stable for any `v1alpha3` cluster and would only need minimal changes to be stable for `v1alpha4`.
+
+### Cluster-api base files
+
+We need to default a big set of custom resources with `cluster-api`.
+These custom resources need to be created and defaulted in a consistent way.
+The approach we follow here is to have a single source of truth which is as complete as possible.
+
+`base files` are our static defaulting values - any webhook or tool which does defaulting should use them as the source of truth.
+
+Let's look at an example to clarify the technical implications.
+A valid `MachinePool` CR in `v1alpha3` can look like this:
+```yaml
+apiVersion: cluster.x-k8s.io/v1alpha3
+kind: MachinePool
+metadata:
+  name: capa-mp-0
+spec:
+  clusterName: capa
+  replicas: 2
+  template:
+    spec:
+      bootstrap:
+        configRef:
+          apiVersion: bootstrap.cluster.x-k8s.io/v1alpha3
+          kind: KubeadmConfig
+          name: capa-mp-0
+      clusterName: capa
+      infrastructureRef:
+        apiVersion: infrastructure.cluster.x-k8s.io/v1alpha3
+        kind: AWSMachinePool
+        name: capa-mp-0
+      version: v1.16.8
+```
+From this custom resource we can now extract the values which are static and configurable.
+So values which a customer might want to change the default of _and_ which we can default statically independent of any other content in the custom resource.
+
+In this particular case there are two values which are both static and configurable:
+- `spec.replicas`
+- `spec.template.spec.version`
+
+We can now construct a `ConfigMap` with these values which are then going to be used by any defaulting tool/webhook:
+```yaml
+apiVersion: v1
+kind: ConfigMap
+metadata:
+  name: machinepool-21-0-0
+  namespace: giantswarm
+data:
+  spec:
+    replicas: 3
+    template:
+      spec:
+        version: v1.21.4
+```
+It is important to note that a assumption is that **no** values in the base file are immutable!
+
+### Cluster-api context dependent values
+
+We have some cases in `cluster-api` where a value in a custom resource is neither part of the fundamental structure nor is it easy to statically default.
+
+These values are usually highly dependent on the context in which the cluster is created in.
+For example the `CIDR` in which the cluster is created.
+We currently assume that these values will be defaulted by custom webhooks or operators with significant provider specific domain knowledge.
```

Comment:
I guess this means we keep them out of scope of this RFC?

Reply:
Yes - mainly mentioned it here for completeness as it was discussed before.